### ISSUE TITLE
Main branch has a test failure on cdash

### ISSUE THREAD

**Issue by ambrad** [E3SM-Project/scream#485]
This [cdash error](https://my.cdash.org/viewTest.php?onlyfailed&buildid=1862147) shows that the main branch has a broken test. The 7-27 cdash doesn't have the error, so it started sometime in the last week.

**Comment by PeterCaldwell** [comment]
Thanks for the catch, Andrew. The fail is "nevapr > tol 1 times. Max rel diff= 5.77548e-08 normalized by ref impl val=1.96844e-06." on just the last timestep in p3_run_and_cmp_cxx_ut_np1_omp1 when log_predictnc=.true. and run on white (GPU). Interestingly, the other 5 timesteps, log_predictnc=.false., other variables, and other machines (do we run nightlies on our other test machines?) didn't have this fail. It also didn't fail in the F90 version of p3_run_and_cmp...

I haven't run our AT test suite yet to confirm I can reproduce this fail (and might not get time to this weekend), but that would be the obvious next step if someone has bandwidth to do it. I did note that nevapr is assigned in p3_evaporate_sublimate_precip and checked the last time we changed that file... which was 15 days ago. So the source of this error isn't obvious.

**Comment by PeterCaldwell** [comment]
The fail wasn't a fluke - it failed in the last 3 nightlies. It started between after early-morning on the 27th and early morning on the 29th. Also, I didn't notice last time that it is *single precision* that is failing. 

Looking at git log, PRs that went in during this time are (in order of likelihood):

-my updating nc and nr after dsd2
-Jeff's p3_main structs PR
- my breaking long lines in F90
-some updates to LLNL machine files
- Aaron's scorpio MPI stuff
- Ben's RRTMGP interface stuff
- Jim's "cime hot fix"

**Comment by jeff-cohere** [comment]
So the failure only occurs in the C++ version of P3, and only in single precision. @PeterCaldwell , you merged my structs PR (#477), right? Since that PR was non-BFB and required new baseline files, did you have to regenerate those baselines somehow? Is it possible that we didn't regenerate the single-precision baseline file? Sorry if this is a dumb question--I don't know the procedures for merging non-BFB changes.

I didn't touch any of the code that calculates anything in P3 in that PR, but I did remove some entries from the baseline files for outputs we eliminated from the top-level P3 interface. This confused the testing system because it assumes all of those entries are in a blessed order. On our last call, we briefly discussed a way of removing this order sensitivity from the baseline files (and I can create an issue that describes that approach if we're interested in doing it).

**Comment by PeterCaldwell** [comment]
@jeff-cohere - whenever you run the AT (which uses gather-all-data), it automatically checks out master and creates a baseline... so you don't need to do it yourself. I'm fairly confident the problem isn't from the structs PR. 

My current hypothetis is that https://github.com/E3SM-Project/scream/blob/debfdf1f5e2b2670d6d22441f46a3c6e9ddd3535/components/scream/src/physics/p3/p3_rain_sed_impl.hpp#L119 might have slightly different conditions for its mask relative to what F90 is doing. I just ran p3_run_and_cmp direct from master at single precision on the GPU machine I have access to and it passed... which makes it harder for me to debug the problem.

**Comment by bartgol** [comment]
If no PR was tested after the last non-BFB PR went in (by means of "admins merge-superpowers"), then the baselines on the SNL filesystem are _not_ up to date. However, I don't think this was the case.

**Comment by jgfouca** [comment]
@PeterCaldwell : I've bisected and confirmed that this PR caused the problem:
```
https://github.com/E3SM-Project/scream/pull/479
```

**Comment by jgfouca** [comment]
Looking at the PR, I really don't see any problems. I can do a "regional bisect" (as in, code regions).

**Comment by ambrad** [comment]
@PeterCaldwell @jgfouca there's a bit of strangeness in the F90 that could be the cause. Look at this line: https://github.com/E3SM-Project/scream/blob/8352f4d5ca4ad5a9dea81ab90428b6e371b1150d/components/cam/src/physics/cam/micro_p3.F90#L3611
It's setting _all_ of nr_incld instead of just nr_incld(k). So it's doing a lot of redundant work (bad) but also violating the qr_smal mass (not BFB).

**Comment by PeterCaldwell** [comment]
Ack. I totally didn't notice that but I agree that it would cause problems. Thanks for the catch. I'll create a PR.

**Comment by singhbalwinder** [comment]
Thats a great find. @PeterCaldwell : If you think this will fix this issue, I will not issue a PR to remove the failing tests as then we have to issue another PR to add them back.

**Comment by PeterCaldwell** [comment]
Yeah, why don't you wait until tomorrow AM before creating your bypass.

**Comment by jgfouca** [comment]
With @ambrad 's change, the fortran remains BFB and the CXX remains non-BFB. My fix PR will keep this change but I will need to find the source of the non-bfbness.

**Comment by PeterCaldwell** [comment]
Yeah, I was just about to report the same thing. On the positive side, I can now reproduce the problem on Lassen (apparently I hadn't pulled before my previous test). @jgfouca - can you dissect the offending PR or should I?

**Comment by jgfouca** [comment]
@PeterCaldwell , I can try.

**Comment by PeterCaldwell** [comment]
Ok, thanks.

**Comment by jgfouca** [comment]
I've been looking at this for a while and haven't found the problem. I'm getting sleepy so I'm going to call it a day and resume tomorrow. Here's what I've found so far:

1) If you take all the changes from #479 and remove the lines that change nc and nr in cloud_sedimentation and rain_sedimentation (3 places in f90, 2 places in cxx). Then the test failure goes away.
2) If you then add back those lines in rain_sedimentation, the test failure comes back.
3) The setting of nr itself appears to be BFB, both
```
nr(k) = nr_incld(k)*rcldm(k)
```
and
```
nr(pk).set(qr_gt_small, nr_incld(pk)*rcldm(pk));
```
Are setting the exact same value for the same indices.
4) The problem appears to be a slight non-bfbness in qrevp in p3_main_part2 at these lines:
```
      nevapr(k)  = qisub + qrevp
      prer_evap(k) = qrevp
      vap_ice_exchange(k) = qidep - qisub + qinuc
      vap_liq_exchange(k) = - qrevp
```
... which is causing nevapr, prer_evap, and vap_liq_exchange to be non-bfb in just one place, in one column, in the very last timestep.
5) What is strange is that these values are set before rain_sedimentation is called, so I don't understand how changing rain_sed could impact these values.

**Comment by ambrad** [comment]
Re: 5, I believe nr in earlier top-level iterations will affect these values.

**Comment by singhbalwinder** [comment]
Since this has not been resolved yet. Do you guys think I should put in a PR to temporarily bypass the failing comparisons?

**Comment by ambrad** [comment]
@singhbalwinder, Jim has a fix in PR #490

**Comment by singhbalwinder** [comment]
Great! I added that `tanh` and was going to add `bfb_tanh` via  PR #469. It didn't break anything when I first added it so I thought I will fix it in a subsequent PR. Sorry about that. I will remove this change from #469.
